### PR TITLE
Add Composition coord map

### PR DESCRIPTION
**Pull request by nilsvu**
## Proposed changes

This makes Jacobians available in Python, which in turn will allow us to take derivatives and integrals over volume data (edit: deferred to next PR).

We can later consider replacing ElementMap with the Composition coordinate map that I'm adding here.

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### CODE REVIEW

File: src/Domain/CoordinateMaps/Composition.hpp [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+ *   domain::element_to_block_logical_map), then deforms the logical cube to a
+ *   wedge using the map described above (Grid frame), and then rotates the grid
+ *   with a time-dependent rotation map (Inertial frame).
+ *
+ * \warning Think about performance implications before using this Composition
+ * class. In an evolution it's usually advantageous to keep at least some of the
+ * maps separate, e.g. to avoid reevaluating the static maps in every time step.
+ * You can also access individual components of the composition in this class.
+ */
+template <typename Frames, size_t Dim,
```
Comment (nilsdeppe) [inline src/Domain/CoordinateMaps/Composition.hpp:62]: can you document the `Frames` template parameter?

Is it possible to write deduction guidelines for the constructor? That would be extremely handy for a class like this...

File: tests/Unit/Domain/CoordinateMaps/Test_Composition.cpp [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+  CHECK_FALSE(map.is_identity());
+  CHECK_FALSE(map.inv_jacobian_is_time_dependent());
+  CHECK_FALSE(map.jacobian_is_time_dependent());
+
+  const auto xi =
+      make_with_random_values<tnsr::I<DataVector, 2, Frame::ElementLogical>>(
+          make_not_null(&generator), make_not_null(&logical_dist),
+          DataVector(5));
+  const auto x = map(xi);
+  const auto jacobian = map.jacobian(xi);
```
Comment (nilsdeppe) [inline tests/Unit/Domain/CoordinateMaps/Test_Composition.cpp:76]: given the non-trivial math being done in the implementation, can you test a 3d map with a full jacobian? At least that jac * inv_jac == identity. Seems tough to get that property to hold by luck :)

**Review by nilsdeppe — CHANGES_REQUESTED** [review]
you can squash immediately :)

File: src/Domain/CoordinateMaps/Composition.hpp (hunk 1 above)
Comment (nilsvu) [reply src/Domain/CoordinateMaps/Composition.hpp:62]: deduction seems to work 👍 see fixup

File: tests/Unit/Domain/CoordinateMaps/Test_Composition.cpp [hunk 3]
```diff
@@ -39,13 +39,12 @@ void test_composition() {
...
   const ElementId<2> element_id{0, {{{1, 0}, {0, 0}}}};
-  const Composition<
-      tmpl::list<Frame::ElementLogical, Frame::BlockLogical, Frame::Inertial>,
-      2>
-      map{element_to_block_logical_map(element_id),
-          std::make_unique<
-              CoordinateMap<Frame::BlockLogical, Frame::Inertial, Affine2D>>(
-              Affine2D{{-1., 1., 0., 1.}, {-1., 1., 1., 3.}})};
+  // Testing template deduction here
+  const Composition map{
```
Comment (nilsdeppe) [inline tests/Unit/Domain/CoordinateMaps/Test_Composition.cpp:45]: Nice! This is a lot simpler :+1:

**Review by nilsdeppe — COMMENTED** [review]
LGTM, please squash!

**Review by nilsdeppe — APPROVED** [review]

**Pull request merged**